### ISSUE TITLE
Improve thread pooling capabilities for Kitura

### ISSUE THREAD

**Issue by rolivieri** [IBM-Swift/Kitura#121]
Kitura should be able to handle many multiple requests arriving concurrently. Think of Java JEE servers for instance. They have a large pool of threads for handling incoming client connections.

**Comment by rolivieri** [comment]
After talking to Daniel, I understand now that our code creates a new dispatch queue task for the processing of each incoming request.

This work item/task is for determining if we could improve the performance of logic/code that executes before the creation of the dispatch queue task.

**Comment by rfdickerson** [comment]
Ok, so let me get this straight- this item is about improving the time between when the client request comes in and when we have the request processed by a thread?

**Comment by rolivieri** [comment]
@rfdickerson Let's talk about this tomorrow in person since we'd like to include as part of this work item the system level threads you and I were talking about for handling i/o operations.

**Comment by shmuelk** [comment]
There is very little that happens between when the request comes in and when the request is placed on the parallel client request queue.

Look at HttpServerSpi.swift lines 50-54 and HttpServer.swift lines 142-145

That's all of the code that runs between the time we get a client socket from BlueSocket until we are running in the block in the client queue.

**Comment by rolivieri** [comment]
@rfdickerson I remembered later today that we had already opened a separate issue to explore the possibility of epoll and kqueue (issue #122). Given this and given that the code (as it is now) creates a new dispatch queue task for the processing of each incoming request, and (as shmuelk’s pointed above) since there is not much happening between the moment a request comes in and the moment the request is placed on the dispatch queue, I am closing this issue. If we should discuss more, let’s reopen and revisit the issue.

**Comment by rfdickerson** [comment]
I think we should draw out the design on the board, I'm not sure I'm understanding the whole mechanism. I created this issue regarding the overhead of spawning and destroying of worker threads.

**Comment by Chatatata** [comment]
+1

**Comment by nmvijay** [comment]
@rfdickerson , I agree with @rolivieri, rather than creating and destroying HTTP request handlers (tasks) for each incoming request, we should have thread pool (similar to database connection pool) which can process the requests concurrently, not in FIFO queue. And this pool should be configurable

**Comment by shmuelk** [comment]
Kitura-net doesn't use threads to handle requests in parallel.

Kitura-net uses Grand Central Dispath (GCD) to enqueue closures onto parallel queues. The HTTPServerRequest and HTTPServerResponse classes in no way represent threads. All of the multi-threading is at the GCD level and is managed by GCD. 

The spawning and handling of worker threads is done automatically by GCD. In fact there are no external controls in GCD that I know of in this area.

**Comment by simplepanda** [comment]
I'm just reading through some of the back discussion re: epoll and kqueue and a comment on #122 as to if dispatch has async assistant mechanisms.

I'm wondering if you fine folks investigated the dispatch async side?

Example:

```swift
   // assume "socket.socketfd" is a socket that's bind()'d to a port and 
   // listen()'d and has non-blocking enabled.

   let queue = dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0)
   let createListener = dispatch_source_create(DISPATCH_SOURCE_TYPE_READ, UInt(socket.socketfd), 0, queue)

   dispatch_source_set_event_handler(listener) {
      let client_fd : dispatch_fd_t = Darwin.accept(socket.socketfd, nil, nil)

      // handle newly received connection, likely by passing it to another
      // dispatch event handler for reading in async fashion
   }
```

Dispatch also has some really nifty stuff with channel reading as far as being able to set low and high "water marks" for how little or much data needs to be in a file descriptor queue before the async reader blocks will enqueue (```dispatch_io_create ```, ```dispatch_io_set_low_water```, ```dispatch_io_set_high_water```, etc).

If you've already been down this road just ignore me but I skimmed the back issues and didn't see mention. Maybe something worth prototyping in a fork for experimentation?

**Comment by shmuelk** [comment]
@simplepanda Actually there are two experimental branches of Kitura-net that implement HTTP keep alive and asynchronous socket I/O. These branches are asynchronous and gcd_asynch.

In both of these branches the accept is done synchronously, while all of the reading is done asynchronously. The branch gcd_asynch uses dispatch_io for all of the I/O on the client socket on both OSX and Linux. The branch asynchronous uses dispatch_io for all of the I/O on the client socket on OSX. On Linux it uses epoll. 

Both of these branches perform about the same on OSX, with a significant improvement there. On Linux the gcd_asynch branch performs worse than the current code without keep alive. The asynchronous branch performs between 70% to 100% faster on Linux.

Various performance people within IBM are involved in understanding what is going on.

**Comment by cloutiertyler** [comment]
What is the status of the gcd async stuff? I'm just now realizing that I have to make all of my requests handling thread safe which could be largely avoided with an asynchronous architecture.

**Comment by shmuelk** [comment]
The old "GCD ASYNCH stuff" has largely been merged into the main kitura-net (the Kitura networking layer) code. On OSX GCD is used and on Linux epoll with non-blocking sockets is used. The reason for not using GCD on Linux, is that the code that uses epoll performs significantly better than the GCD code.

**Comment by cloutiertyler** [comment]
Thanks for the update! And that is good to know about epoll, I didn't realize their would be a significant performance difference for GCD on different platforms.

As far as I can tell, it looks like Kitura-Net is still using a block accept call in which each client connection is accepted on a different thread. 

https://github.com/IBM-Swift/Kitura-net/blob/master/Sources/KituraNet/HTTP/HTTPServer.swift#L196

Can you point me to the non-blocking code?

**Comment by cloutiertyler** [comment]
Essentially, the thread management is killing me with Kitura since my in memory server state (necessary for testing) is having data races all over the place. I created https://github.com/skylab-inc/Edge which implements a single threaded GCD non-blocking API, albeit crudely. I was hoping that perhaps Kitura was intending to move in that direction.

**Comment by shmuelk** [comment]
You are correct in that the accept is done in a blocking thread. However, the reading, parsing, and processing of the requests is done in blocks queued on the Global Queue.

See https://github.com/IBM-Swift/Kitura-net/blob/fc6e2178bba46b3209709c2e331b3ad4c402e780/Sources/KituraNet/HTTP/HTTPServer.swift#L193
